<a href="https://colab.research.google.com/github/armandoordonez/eda_couse/blob/main/Merge_de_datasets.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Practica de Merge de Datasets

## 0.Importar Librerias
Iniciamos importando las librerias requeridas para esta practica

In [3]:
import pandas as pd
import numpy as np
import seaborn as sns

sns.set_theme(style="whitegrid")

## 1.Cargar los datasets
Lo primero que debemos hacer es cargar los datasets. Esto lo podemos hacer a través de Drive o subiendo el archivo al colab

In [ ]:
#Importamos la libreria de drive

from google.colab import drive
drive.mount('/gdrive')

Drive already mounted at /gdrive; to attempt to forcibly remount, call drive.mount("/gdrive", force_remount=True).


In [7]:
# Opción Google drive
# Usamos pandas para leer el csv en la ruta de nuestro drive
#df1 = pd.read_csv('/gdrive/MyDrive/Monitoria IA/Practicas/Merge/Personas1.csv',    #Se ingresa la ruta completa a la ubicación del archivo de interés. El MyDrive de la ruta es tu "Mi Unidad", desde ahí vas ingresando a cada carpeta que tengas hasta llegar a tu archivo.
 #                na_values='?')              #Como el archivo contiene algunos caracteres ?, con este comando pandas los interpreta como nan (not a number).

#df2 = pd.read_csv('/gdrive/MyDrive/Monitoria IA/Practicas/Merge/Personas2.csv',
#                 na_values='?')

In [8]:
# Opción GITHUB
#df1 = pd.read_csv('https://raw.githubusercontent.com/armandoordonez/eda_couse/main/data/Personas1.csv',    #Se ingresa la ruta completa a la ubicación del archivo de interés. El MyDrive de la ruta es tu "Mi Unidad", desde ahí vas ingresando a cada carpeta que tengas hasta llegar a tu archivo.
#                na_values='?')              #Como el archivo contiene algunos caracteres ?, con este comando pandas los interpreta como nan (not a number).

#df2 = pd.read_csv('https://raw.githubusercontent.com/armandoordonez/eda_couse/main/data/Personas1.csv',na_values='?')

### Subir archivo a Colab
Para subir el archivo a colab, damos click en el icono de carpeta que está en el menú del lado izquierdo y luego presionamos el icono del archivo con una flecha hacia arriba.
Esto nos abrirá unan ventana con el explorador de archivos, donde podremos importar el dataset.
Finalmente, solo se importa usando el nombre del csv y definiendo cuales valores interpretará pandas como nan.

In [9]:
#df1 = pd.read_csv('./Personas1.csv',na_values='?')
#df2 = pd.read_csv('./Personas2.csv', na_values='?')

## 2. Verificamos los DF

In [10]:
df1.head()

first_name   last_name    cc
0     Berget   Proughten  1495
1     Scotti  Yegorovnin   311
2    Clayton       Reide   914
3     Toddie     Haddick   671
4     Brynna    Ferrucci   692

Podemos ver que el primero consta de una lista de personas que tienen un nombre y apellido, así como el número de cedula.

In [11]:
df2.head()

first_name   last_name    cc
0     Berget   Proughten  1495
1     Scotti  Yegorovnin   311
2    Clayton       Reide   914
3     Toddie     Haddick   671
4     Brynna    Ferrucci   692

El segundo df consta del salario, la ciudad y el número de cedula de una persona, solo que en este caso desconocemos el nombre.

## 3. Problema del merge
Si quisieramos poder saber el nombre y apellido de la persona que está ganando ese salario y vive en esa ciudad, ¿Cómo lo haríamos?


Para problemas como este, en el que tenemos dos o más datasets que necesitamos "pegar" tomando como referencia sus columnas, utilizamos la función merge que nos brinda Pandas.

In [12]:
merged_df = pd.merge(df1, df2, on='cc')

Una posible solución es la linea de código que está arriba, en la cual tomamos ambos df y los pegamos tomando como referencia los números de cedula.

In [13]:
merged_df.head()

first_name_x last_name_x    cc first_name_y last_name_y
0       Berget   Proughten  1495       Berget   Proughten
1       Scotti  Yegorovnin   311       Scotti  Yegorovnin
2       Scotti  Yegorovnin   311       Arlena  Raubenheim
3       Scotti  Yegorovnin   311        Brana       Obell
4       Arlena  Raubenheim   311       Scotti  Yegorovnin

Como podemos observar, el resultado nos muestra que tenemos una nueva tabla con nombre, apellido, numero de cedula, salario y ciudad, logrando así unir los datos.
En este caso, el df resultante puede no tener el mismo tamaño que los df iniciales, esto sería si hay cedulas que no coinciden en ambos dataset.

## Tipos de Merge

Los principales tipos de merge se pueden separar en 3:
*   Inner: Es el tipo de merge por defecto, solo toma los datos que coincidan en su columna de referencia para ambos dataset. (En el ejemplo fue la cc, si en alguno de los dos dataset no estaba la misma cc, no se agrega)
*   Left: Agrega los datos que estén en ambos dataset, pero también agrega los datos del dataset que pongamos primero en nuestra función merge. (En el ejemplo, el df1)
*   Right: Agrega los datos que estén en ambos dataset, pero también agrega los datos del dataset que pongamos primero en nuestra función merge. (En el ejemplo, el df1)






Para cambiar el tipo de Merge debemos añadir una instrucción a la función llamada "how":


```
pd.merge(df1,df2,on='cc',how="left")
```



In [14]:
merged_df_left = pd.merge(df1,df2,on='cc', how='left')

In [15]:
merged_df_left.head()

first_name_x last_name_x    cc first_name_y last_name_y
0       Berget   Proughten  1495       Berget   Proughten
1       Scotti  Yegorovnin   311       Scotti  Yegorovnin
2       Scotti  Yegorovnin   311       Arlena  Raubenheim
3       Scotti  Yegorovnin   311        Brana       Obell
4      Clayton       Reide   914      Clayton       Reide

Se puede apreciar que, por ejemplo, para los dos primeros datos, falta el salario y la ciudad. Esto es porque al ser de tipo left, no encontró la cedula 1495 en el df2, por lo que no pudo emparejarle un salario o ciudad, lo que deja esos datos en NaN.

In [16]:
merged_df_right = pd.merge(df1,df2,on='cc', how='right')

In [17]:
merged_df_right.head()

first_name_x last_name_x    cc first_name_y last_name_y
0       Berget   Proughten  1495       Berget   Proughten
1       Scotti  Yegorovnin   311       Scotti  Yegorovnin
2       Arlena  Raubenheim   311       Scotti  Yegorovnin
3        Brana       Obell   311       Scotti  Yegorovnin
4      Clayton       Reide   914      Clayton       Reide

En el caso del right, pasa algo similar pero en caso contrario, puesto que no encontró para el primero caso la cedula 611 en el df1, por lo que deja el nombre y apellido en NaN, así también para otros casos.

## Nota:
También es posible cambiar las columnas que se toman de referencia al momento de hacer merge.


```
pd.merge(df1,df2,on=['cc','city'])
```
Podemos tener una lista de columnas y al momento de hacer el merge, si lo tenemos en su forma estándar (inner), solo traerá y emparejará los datos que tenga la misma cedula y ciudad en ambos datasets.
